flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/sink/SinkUpsertMaterializer.java

```diff
@@ -96,36 +98,39 @@ public void open() throws Exception {
 
     @Override
     public void processElement(StreamRecord<RowData> element) throws Exception {
-        RowData row = element.getValue();
-        boolean isInsertOp = row.getRowKind() == INSERT || row.getRowKind() == UPDATE_AFTER;
-        // Always set the RowKind to INSERT, so that we can compare rows correctly (RowKind will
-        // be ignored)
-        row.setRowKind(INSERT);
+        final RowData row = element.getValue();
         List<RowData> values = state.value();
         if (values == null) {
             values = new ArrayList<>(2);
         }
 
-        if (isInsertOp) {
-            values.add(row);
-            // Update to this new one
-            collector.collect(row);
-        } else {
-            int lastIndex = values.size() - 1;
-            int index = removeFirst(values, row);
-            if (index == -1) {
-                LOG.info(STATE_CLEARED_WARN_MSG);
-                return;
-            }
-            if (values.isEmpty()) {
-                // Delete this row
-                row.setRowKind(DELETE);
+        switch (row.getRowKind()) {
+            case INSERT:
+            case UPDATE_AFTER:
+                row.setRowKind(values.isEmpty() ? INSERT : UPDATE_AFTER);
+                values.add(row);
```

Comment:
Can we still keep update an insert row? This will make understanding logic easier. So we don't need modify equals in `removeFirst`.

Reply:
Maybe I should have added a comment. This is a performance optimization. If we would set INSERT for the state, we would need to copy the row to emit `UPDATE_AFTER`. This implementation prevents unnecessary copying by simply setting the row kind when necessary.

Reply:
I get your point, if the state backend is HeapBackend instead of RocksDb.
Can we collect first and then update state?

Reply:
Not sure if this is safe in case of object reuse. Modifying an emitted record sounds dangerous to me, it could well be that the following operator puts the row in state. Or (as we do as well in the input) modify the row kind.

Reply:
I think this is OK to reuse output object. For example, in the CalcOperator generated by StreamExecCalc, it always reuse output `RowData` object.

On the other hand, can we copy the data if HeapStateBackend? For example, 
when `StateConfigUtil.isStateImmutableInStateBackend` return false, just copy the record.

Reply:
I will update the PR. But checking the state backend and copying should definitely be avoided. The current solution is not super bad if we document that the row kind can be arbitrary in state.

Reply:
OK, let's document it.